sirepo/auth_db.py

```diff
@@ -131,27 +130,77 @@ def delete_roles(cls, uid, roles):
                     cls.role.in_(roles),
                 ).delete(synchronize_session='fetch')
                 cls._session.commit()
+                audit_proprietary_lib_files(uid)
+
+        @classmethod
+        def has_role(cls, uid, role):
+            with thread_lock:
+                return bool(cls.search_by(uid=uid, role=role))
 
         @classmethod
         def uids_of_paid_users(cls):
+            import sirepo.auth
+
             return [
                 x[0] for x in cls._session.query(cls).with_entities(cls.uid).filter(
                     cls.role.in_(sirepo.auth.PAID_USER_ROLES),
                 ).distinct().all()
             ]
 
+
     # only creates tables that don't already exist
     UserDbBase.metadata.create_all(_engine)
 
 
+def audit_proprietary_lib_files(uid):
+    """Add/removes proprietary files based on a user's roles
+
+    For example, add the Flash rpm if user has the flash role.
+
+    Args:
+        uid: user to audit
+    """
+    import py
+    import pykern.pkio
+    import sirepo.auth
+    import sirepo.feature_config
+    import sirepo.simulation_db
+    import sirepo.srdb
+
+    x = sirepo.feature_config.cfg().proprietary_sim_types
+    if not x:
+        return
+    for t in x:
+        r = UserRole.has_role(
+            uid,
+            sirepo.auth.role_for_sim_type(t),
+        )
+        for f in pykern.pkio.sorted_glob(
+            sirepo.srdb.proprietary_code_dir(t).join('*'),
+        ):
+#TODO(robnagler) ensure no collision on names with uploaded files
+# (restrict suffixes in user uploads)
+            p = sirepo.simulation_db.simulation_lib_dir(t, uid=uid).join(f.basename)
```

Comment:
Why not use sirepo.auth.set_user() instead of passing in the uid?

Reply:
#2588 discusses this in detail. set_user is global state, which is bad, because this is local context: simulation_lib_dir for uid. Also set_user requires that flask context be configured, because it sets the cookie (or set_cookie_for_utils) which is too much coupling here. We aren't logging in the user here; we are performing a database operation in an unspecified context.